packages/shared/lib/services/connection.service.ts

```diff
@@ -578,39 +585,43 @@ class ConnectionService {
         const credentials = connection.credentials as OAuth2Credentials;
         const providerConfigKey = connection.provider_config_key;
 
-        const runningRefresh = this.findRunningRefresh(connectionId, providerConfigKey);
-        if (runningRefresh) {
-            return runningRefresh.promise;
-        }
-
         const shouldRefresh = await this.shouldRefreshCredentials(connection, credentials, providerConfig, template, instantRefresh);
 
         if (shouldRefresh) {
+            // We must ensure that only one refresh is running at a time accross all instances.
+            // Using a simple redis entry as a lock with a TTL to ensure it is always released.
+            // NOTES:
+            // - This is not a distributed lock and will not work in a multi-redis environment.
+            // - It could also be unsafe in case of a Redis crash.
+            // We are using this for now as it is a simple solution that should work for most cases.
+            const lockKey = `lock:refresh:${environment_id}:${providerConfigKey}:${connectionId}`;
             try {
+                const ttlInMs = 10000;
+                const acquitistionTimeoutMs = ttlInMs * 1.2; // giving some extra time for the refresh to complete
```

Comment:
How was this `1.2` time determined? Is that enough time?

Reply:
in theory acquisition time could be the same as ttl but this is just adding some extra padding and allow the lock acquisition to eventually succeed even when 2 acquisition happens at the very same time.

Reply:
maybe the comment is confusing actually. it is not giving extra time for the refresh but extra time given to the acquisition in case the initial acquirer is timing out
Let me change the comment